devices/AtomQrCode/QrCodeReader.cs

```diff
@@ -0,0 +1,843 @@
+        }
+
+        /// <summary>
+        /// Set the flood light mode.
+        /// </summary>
+        public FloodLight FloodLight
+        {
+            set
+            {
+                byte[] suffix = new byte[] { 0x02, 0x00, 0xFE, 0x32 };
+
+                switch (value)
+                {
+                    case FloodLight.OnWhenReading:
+                        // this is the default on the suffix
+                        break;
+
+                    case FloodLight.AlwaysOn:
+                        suffix[_param1Index] = 0x01;
+                        suffix[_param2Index] = 0x31;
+                        break;
+
+                    case FloodLight.AlwaysOff:
+                        suffix[_param1Index] = 0x02;
+                        suffix[_param2Index] = 0x30;
+                        break;
+                }
+
+                SendCommandType08(suffix);
+            }
+        }
+
+        /// <summary>
+        /// Set the terminator to be appended to the code data.
+        /// </summary>
+        public Terminator Terminator { set; private get; } = Terminator.None;
+
+        /// <summary>
+        /// Creates a QR Code Reader.
+        /// </summary>
+        /// <param name="portName">The port to use (for example, COM1).</param>
+        public QrCodeReader(string portName)
+        {
+            _portName = portName;
+
+            _disposed = false;
+
+            // instantiate serial port with reader default baud rate
+            _readerSerialPort = new SerialPort(portName, baudRate: 9600);
+
+            // setup event handler
+            _readerSerialPort.DataReceived += Reader_DataReceived;
+
+            // set terminator and "new line" as TAB
+            _readerSerialPort.WatchChar = '\t';
+            _readerSerialPort.NewLine = "\t";
+
+            // open serial port
+            _readerSerialPort.Open();
+
+            // setup buffers
+            _commandBuffer = new byte[11];
+            _readBuffer = new byte[256];
+
+            PerformInitialConfig();
+        }
+
+        /// <summary>
+        /// Start scanning for barcodes in automatic mode.
+        /// Requires to set the scan mode to <see cref="TriggerMode.Autosensing"/>.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">If <see cref="TriggerMode"/> hasn't been set to <see cref="TriggerMode.Continuous"/> previously.</exception>
+        public void StartScanning()
+        {
+            if (TriggerMode != TriggerMode.Autosensing)
+            {
+                throw new InvalidOperationException();
+            }
+
+            // setup thread to process barcode reading
+            new Thread(() =>
+            {
+                // hang in here to let the command flow to make sure the thread doesn't exit immediately
+                Thread.Sleep(100);
+
+                do
+                {
+                    if (_newDataAvailable.WaitOne((int)_lastTimeout.TotalMilliseconds, true))
```

Comment:
The WaitOne is cancelled only after the timeout. That is blocking here even if the _waitingForData if true.
In the mean time, many things can happen!

Reply:
But I may be wrong and it may all goes nicely. But for those caces, to keep it simpler than the CancellationToken, a simple check that you exited the thread with a bool as well can make it. And you can only start it again if that boolean is reset